SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl

```diff
@@ -1276,6 +1276,39 @@ void MechanicalObject<DataTypes>::reinit()
 
     if (translation.getValue()[0]!=0.0 || translation.getValue()[1]!=0.0 || translation.getValue()[2]!=0.0)
         this->applyTranslation( translation.getValue()[0],translation.getValue()[1],translation.getValue()[2]);
+
+    if (x.getValue()[0].size() == 7)   // The object is a Rigid3D.
```

Comment:
To answer your question, as this is templated code the right approach is probably to rely on code specialization instead of type branching. 

Eg:
```cpp
template<A>
MechanicalObject<A>::reinit()
{
     commonInitAllDataType();
     specificInitDependingOnData();
}

template<class A>
MechanicalObject<A>::commonInitAllDataType() { ....blahblha... }

template<class A>
MechanicalObject<A>::specificInitDependingOnData() {} /// Nothing by default
  
template<> /// Specialization on rigid2
MechanicalObject<Rigid3D>::specificInitDependingOnData() 
{
      ///put here the code related to quaternion normalization
} 
```